posts/aws-cognito-and-jhipster-for-the-love-of-oauth-2.0.md

```diff
@@ -0,0 +1,86 @@
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/cognito-jhipster/images/2020/02/cognito-app-client.png)
+
+The app client id and secret are required by the JHipster app, make sure to not share them or push them to a public repository. And also, keep in mind that secrets cannot be changed after an app is created. You can create a new app with a new secret if you want to rotate the secret that you are using.
+
+The callback, sign out URL and the allowed OAuth Flows/Scopes need to have the values below:
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/cognito-jhipster/images/2020/02/cognito-app-client-settings.png)
+
+Finally, a domain is required in order to have the login flow working. You can configure the domain in the **Domain name** section (you can create or use your own domain).
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/cognito-jhipster/images/2020/02/cognito-domain.png)
+
+# Code changes
+
+## Secrets
+
+The first thing to do is to replace the `client-id` and `client-secret` with the correct values in `src/main/resources/config/application.yml`. Those secrets can be viewed in the **App clients** section:
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/cognito-jhipster/images/2020/02/cognito-secrets.png)
+
+Then the `issuer-uri` must be updated with the correct URI: `https://cognito-idp.REGION.amazonaws.com/POOL_ID` (the pool id can be viewed in the **General settings** section).
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/cognito-jhipster/images/2020/02/cognito-pool.png)
+
+## Roles
+
+The way JHipster extracts the roles from the token doesn't work for Cognito since they are stored in the `cognito:groups` attribute. This [commit](https://github.com/Falydoor/cognito-jhipster/commit/e2cceba1f5e9844cb0ab4ca6f7601b1c3a8b96a4) contains the changes that will make the roles extraction works.
+
+## Login and logout fixes
+
+The method `getUser` in `UserService.java` extracts the information from the Cognito's token and save it to the local database. The login value is not set correctly and uses the `id` of the user instead of the email. A simple override with the email value fixes the issues, more details can be found on this [commit](https://github.com/Falydoor/cognito-jhipster/commit/69eae0a890de8401174214e8db3381e21f8e9789).
+
+The last fix is related to the logout since JHipster generates the logout URL for Okta/Keycloak but not for Cognito. The correct URL is formatted like this: `https://DOMAIN.auth.REGION.amazoncognito.com/logout`, this [commit](https://github.com/Falydoor/cognito-jhipster/commit/53acb5c2c53873eafc6a9631ee0fad9630d3ea05) changes how the URL is retrieved (don't forget to replace `DOMAIN` and `REGION` in `src/main/java/com/mycompany/myapp/web/rest/LogoutResource.java`).
+
+# Conclusion
+
+Finally, run your application with `./mvnw` and you should see the Cognito login screen after clicking on `sign in` (make sure the logout flow works as well):
```

Comment:
Once you run your application with `.mvnw`, you should see the Cognito login screen after clicking on `sign in` (ensure the logout flow works properly as well):